### PR TITLE
Add pack_command to support writing via hiredis-py

### PR DESCRIPTION
**Pull request by prokazov**
As of now, hiredis-py only exposes Reader functionality, that helps to speed up the RESP deserialization.
pack_command serializes redis-py command (tuple of  str, bytes, int or float) to a bytes object with the RESP representation of the command.

### CODE REVIEW

File: hiredis/version.py [hunk 1]
```diff
@@ -1 +1 @@
-__version__ = "2.1.1"
+__version__ = "2.1.2"
```
Comment (chayim) [inline hiredis/version.py:1]: This will become 2.2.0 since it's not a minor change. Mind removing from this PR?

File: tests/test_pack.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
+import hiredis
+import pytest
+
+testdata = [
```
Comment (chayim) [inline tests/test_pack.py:4]: very complete. excellent

File: tests/test_pack.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+
+@pytest.mark.parametrize("cmd,expected_packed_cmd", testdata, ids=testdata_ids)
+def test_basic(cmd, expected_packed_cmd):
+    packed_cmd = hiredis.pack_command(cmd)
+    assert packed_cmd == expected_packed_cmd
+
+
+def test_wrong_type():
+    with pytest.raises(TypeError):
```
Comment (chayim) [inline tests/test_pack.py:38]: Good catch given the current implementation

File: src/pack.c [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        PyErr_SetString(PyExc_RuntimeError,
+                        "Failed to serialize the command.");
+        goto cleanup;
+    }
+
+    result = PyBytes_FromStringAndSize(resp_bytes, len);
+    hi_free(resp_bytes);
+cleanup:
+    sdsfreesplitres(tokens, tokens_number);
+    hi_free(lenghts);
```
Comment (chayim) [inline src/pack.c:104]: Let's rename ```lenghts``` to be ```lengths```

**Review by chayim — CHANGES_REQUESTED** [review]
Big fan, very small changes. Any chance you've finished up on the redis-py PR we were discussing that shows covered usage?

File: src/pack.c (hunk 4 above)
Comment (prokazov) [reply src/pack.c:104]: Done.

File: hiredis/version.py (hunk 1 above)
Comment (prokazov) [reply hiredis/version.py]: Done.
Added an entry to the change log.

**Comment by prokazov** [comment]
> Any chance you've finished up on the redis-py PR ...
I'll publish a draft PR soon but as of now:
https://github.com/redis/redis-py/compare/master...prokazov:redis-py:master
It's kind of non-trivial to properly test redis-py against unpublished hiredis-py.

**Comment by prokazov** [comment]
I mean I tested the redis-py with the new pack_command locally, but its CI/CD actions target the stock hiredis-py.
The changes above only cover the regular ("sync") API. The async one would be basically a "copy-paste" of the sync API (with a review feedback). For some reason the async redis-py follows that powerful design pattern.

**Comment by chayim** [comment]
@prokazov Want to tag me and @dvora-h there... we can continue there. This looks good. Once CI reports in, etc, I'll prep a release - for today or Sunday.

**Review by chayim — APPROVED** [review]

File: setup.py [hunk 5]
```diff
@@ -4,53 +4,58 @@
...
+
 ext = Extension("hiredis.hiredis",
-  sources=sorted(glob.glob("src/*.c") +
-                 ["vendor/hiredis/%s.c" % src for src in ("alloc", "read", "sds")]),
-  extra_compile_args=["-std=c99"],
-  include_dirs=["vendor"])
+                sources=sorted(glob.glob("src/*.c") +
+                               ["vendor/hiredis/%s.c" % src for src in ("alloc", "async", "hiredis", "net", "read", "sds")]),
+                extra_compile_args=["-std=c99"],
+                extra_link_args=["-Wl,-Bsymbolic"],
```
Comment (chayim) [inline setup.py:22]: -Bsymbolic fails across many platforms - please see CI

**Review by chayim — CHANGES_REQUESTED** [review]
Changes, as per Bsymbolic build fails

File: setup.py (hunk 5 above)
Comment (prokazov) [reply setup.py]: @chayim @dvora-h 
Removed.
It crashes in RedisGears without the extra link args. Though we can figure this out in a different PR.

Comment (zalmane) [reply setup.py]: I suggest we leave it in there except for the MacOS or platforms that don't support the flag. Or vice versa, but this conflicts with Redis' hiredis and crashes in RE without this

**Comment by zalmane** [comment]
@chayim I pushed a commit with conditional flags. Can you re-run the CI? Are there any wheels built not covered by the CI? I excluded win and darwin platforms which don't support the flag. The CI failure of the ubuntu on 3.7 seemed unrelated.

File: setup.py (hunk 5 above)
Comment (chayim) [reply setup.py]: Bsymbolic is a pretty large hammer to wield. Given that we precompile this and push various wheels to pypi, I'd like to be careful with this outcome, especially as we're now changing everyone.

I think you should be safe with the following ```return ["-std=c99", "-static-libstdc++", "-O2"]``` in place of the current ```-stc=c99``` should acheive the same effect. Note: we did add a compiler optimization there, and one could argue removing and -O3 as well (which I'm open to).

@prokazov what happens with gears (and others) in this case? Symbol collisions or sanity?

**Comment by chayim** [comment]
Looks great. Given a past bug, I just added #148. I'll merge that into your branch, and see it through. If we're good there, we're ready to merge!

**Pull request merged**